src/bin/hal/cmd/psbt.rs

```diff
@@ -429,3 +432,106 @@ fn exec_merge<'a>(matches: &clap::ArgMatches<'a>) {
 		print!("{}", base64::encode(&merged_raw));
 	}
 }
+
+fn cmd_rawsign<'a>() -> clap::App<'a, 'a> {
+	cmd::subcommand("rawsign", "sign a psbt with private key and add sig to partial sigs").args(&[
+		cmd::arg("psbt", "PSBT to finalize, either base64/hex or a file path").required(true),
+		cmd::arg("input-idx", "the input index to edit").required(true),
+		cmd::arg("priv-key", "the private key in WIF/hex").required(true),
+		cmd::arg("compressed", "Whether the corresponding pk is compressed")
+			.required(false)
+			.default_value("true"),
+		cmd::opt("raw-stdout", "output the raw bytes of the result to stdout")
+			.short("r")
+			.required(false),
+		cmd::opt("output", "where to save the resulting PSBT file -- in place if omitted")
+			.short("o")
+			.takes_value(true)
+			.required(false),
+	])
+}
+
+// Get the scriptpubkey/amount for the psbt input
+fn get_spk_amt(psbt: &psbt::PartiallySignedTransaction, index: usize) -> (&bitcoin::Script, u64) {
+    let script_pubkey;
+    let amt;
+	let inp = &psbt.inputs[index];
+    if let Some(ref witness_utxo) = inp.witness_utxo {
+        script_pubkey = &witness_utxo.script_pubkey;
+		amt = witness_utxo.value;
+    } else if let Some(ref non_witness_utxo) = inp.non_witness_utxo {
+        let vout = psbt.global.unsigned_tx.input[index].previous_output.vout;
+        script_pubkey = &non_witness_utxo.output[vout as usize].script_pubkey;
+		amt = non_witness_utxo.output[vout as usize].value;
+    } else {
+		panic!("Psbt missing both witness and non-witness utxo")
+    }
+    (script_pubkey, amt)
+}
+
+fn exec_rawsign<'a>(matches: &clap::ArgMatches<'a>) {
+	let (raw, source) = file_or_raw(&matches.value_of("psbt").unwrap());
+	let mut psbt: psbt::PartiallySignedTransaction = deserialize(&raw).expect("invalid PSBT format");
+
+	let raw = matches.value_of("priv-key").expect("no key provided");
+	let idx = matches.value_of("input-idx").expect("Input index not provided");
+	let i : usize = idx.parse().expect("Expected number");
+	let compressed = matches.value_of("compressed").unwrap().parse::<bool>()
+		.expect("Compressed must be boolean");
+
+	let (spk, amt) = get_spk_amt(&psbt, i);
+	let redeem_script = psbt.inputs[i].redeem_script.as_ref().map(|x|
+		bitcoin::blockdata::script::Builder::new()
+		.push_slice(x.as_bytes())
+		.into_script());
+	let witness_script = psbt.inputs[i].witness_script.as_ref()
+		.map(|x| vec![x.clone().into_bytes()]);
+	let witness = witness_script.unwrap_or(Vec::new());
+	let script_sig = redeem_script.unwrap_or(bitcoin::Script::new());
+
+	// Call with age and height 0.
+	// TODO: Create a method to rust-bitcoin psbt that outputs sighash
+	// Workaround using miniscript interpreter
+	let interp = miniscript::Interpreter::from_txdata(spk, &script_sig, &witness, 0, 0)
+		.expect("Witness/Redeem Script is not a Miniscript");
+	let tx = psbt.clone().extract_tx();
```

Comment:
Is there no other way to get the tx our without closing the entire psbt? (I don't know if the compiler optimizes the pain of this, but isn't there an accessor to the `&Transaction` that you then can clone?

Reply:
I think `psbt.global.unsigned_tx.clone()` should do.

Reply:
Oh you don't even seem to need an owned tx as all, so you can just do `psbt.global.unsigned_tx` later in the `sighash_message` call.